util/include/MultiSizeBufferPool.hpp

```diff
@@ -0,0 +1,483 @@
+#include "diagnostics.h"
+#include "performance_handler.h"
+
+#include <boost/lockfree/queue.hpp>
+
+#include <string_view>
+#include <mutex>
+#include <condition_variable>
+#include <atomic>
+#include <set>
+#include <deque>
+
+/**
+ * Multi-Buffer Memory Pool (MultiSizeBufferPool):
+ *
+ * Definitions:
+ * - Buffer: a continuous, fixed size array of bytes dedicated to a caller.
+ * - Chunk: Wraps a buffer.Pool deals internally with chunks, which include header and trailer metadata for internal
+ * use.
+ * - Used buffer/chunk: currently used by caller
+ * - Unused buffer/chunk: currently not-used by caller, and stored in the pool.
+ * - Allocated buffer/chunk: currently allocated by pool. Can be used or unused in that case.
+ * - Non-allocated buffer/chunk: currently not allocated by pool - pool reserves the right to allocate this chunk
+ * dynamically.
+ * - Subpool: holds a group of buffers of a fixed constant size. Each subpool has its own configuration and contention.
+ * lock-free domain: threads that try to get a buffer from subpool A, do not compete with threads that try to get a
+ * buffer from subpool B. This allow (with an additional very small implementation changes) to create multiple
+ * subpools with the same buffer size to reduce contention. Currently, all subpools should be different sizes.
+ * - Client - a user of the pool's API in order to use the memory pool buffer.
+ * - IPurger: a module that is responsible for purging the sub-queues, based on recently collected data. Can be
+ * disabled. Can be passed externally by pool creator. If not passed externally, the pool creates a default
+ * DefaultPurger. The main idea of the purger is to deallocate unused resources (chunks) from the pool to keep it
+ * minimal according to a given purging policy. Purger can be configured to treat different sub-pools (buffer sizes)
+ * in different ways.
+ * - ISubpoolSelector: a module that is responsible for selecting the most suitable sub-pool (buffer size), based on
+ * recently collected data. Can be disabled. It can be passed externally by the pool creator. If not passed externally,
+ * pool creates a default DefaultSubpoolSelector which chooses to smallest buffer size sub-pool always.
+ * - Pool - holds its configuration purger, subpool selector and an array of subpools (at least 1).
+ *
+ * Overview:
+ * This memory pool has the following attributes:
+ * - Thread safe for all of its operations (not thread safe, of course, when accessing the currently used buffers).
+ * - Supports advanced configuration, in particular caller can define any amount of sub-pools of any size with initial
+ * and maximal number of buffers for each subpool.
+ * A caller may control the overall size of the allocated bytes in the pool, this can give more 'freedom' per subpool,
+ * and restrict the total pool size.
+ * - Rich statistics:
+ * 1) Histograms (snapshots) reported to logs. Currently, histograms are for this pool can't be views
```

Comment:
histograms for this pool can't be viewed

Reply:
Thanks, fixed